Sources/NIOWebSocket/WebSocketFrameDecoder.swift

```diff
@@ -0,0 +1,321 @@
+public enum NIOWebSocketError: Error {
+    /// The frame being sent is larger than the configured maximum
+    /// acceptable frame size
+    case invalidFrameLength
+
+    /// A control frame may not be fragmented.
+    case fragmentedControlFrame
+
+    /// A control frame may not have a length more than 125 bytes.
+    case multiByteControlFrameLength
+}
+
+internal extension WebSocketErrorCode {
+    internal init(_ error: NIOWebSocketError) {
+        switch error {
+        case .invalidFrameLength:
+            self = .messageTooLarge
+        case .fragmentedControlFrame,
+             .multiByteControlFrameLength:
+            self = .protocolError
+        }
+    }
+}
+
+public extension ByteBuffer {
+    /// Applies the WebSocket unmasking operation.
+    ///
+    /// - parameters:
+    ///     - maskingKey: The masking key.
+    public mutating func unmask(_ maskingKey: WebSocketMaskingKey, indexOffset: Int = 0) {
+        /// Shhhh: secretly unmasking and masking are the same operation!
+        mask(maskingKey, indexOffset: indexOffset)
+    }
+
+    /// Applies the websocket masking operation.
+    ///
+    /// - parameters:
+    ///     - maskingKey: The masking key.
+    ///     - indexOffset: An integer offset to apply to the index into the masking key.
+    ///         This is used when masking multiple "contiguous" byte buffers, to ensure that
+    ///         the masking key is applied uniformly to the collection rather than from the
+    ///         start each time.
+    public mutating func mask(_ maskingKey: WebSocketMaskingKey, indexOffset: Int = 0) {
+        self.withUnsafeMutableReadableBytes {
+            for (index, byte) in $0.enumerated() {
+                $0[index] = byte ^ maskingKey[(index + indexOffset) % 4]
+            }
+        }
+    }
+}
+
+/// The current state of the frame decoder.
+enum DecoderState {
+    /// Waiting for a frame.
+    case idle
+
+    /// The initial frame byte has been received, but the length byte
+    /// has not.
+    case firstByteReceived
+
+    /// The length byte indicates that we need to wait for the length word, and we're
+    /// currently waiting for it.
+    case waitingForLengthWord
+
+    /// The length byte indicates that we need to wait for the length qword, and
+    /// we're currently waiting for it.
+    case waitingForLengthQWord
+
+    /// The mask bit indicates we are expecting a mask key.
+    case waitingForMask
+
+    /// All the header data is complete, we are waiting for the application data.
+    case waitingForData
+}
+
+enum ParseResult {
+    case insufficientData
+    case continueParsing
+    case result(WebSocketFrame)
+}
+
+/// An incremental websocket frame parser.
+///
+/// This parser attempts to parse a websocket frame incrementally, keeping as much parsing state around as possible to ensure that
+/// we don't repeatedly partially parse the data.
+struct WSParser {
+    var firstByte: UInt8? = nil
```

Comment:
the `reset` method is private. Shouldn't those be private then too?

Reply:
ok, then `reset` should maybe have the same visibility?

Reply:
Why? `reset` is an action taken by the state machine. How about I make these `internal private(set)` instead?